src/compiler/core.ts

```diff
@@ -1489,45 +1528,264 @@ namespace ts {
+                // upper-case form as some unicode characters do not properly round-trip to
+                // lowercase (such as `ẞ` (German sharp capital s)).
+                return compareCaseSensitive(a.toUpperCase(), b.toUpperCase());
+            }
+
+            function compareCaseSensitive(a: string, b: string) {
+                return a < b ? Comparison.LessThan : a > b ? Comparison.GreaterThan : Comparison.EqualTo;
+            }
+
+            function compareCaseSensitiveDictionaryOrder(a: string, b: string) {
+                // An ordinal comparison puts "A" after "b", but for the UI we want "A" before "b".
+                // We first sort case insensitively.  So "Aaa" will come before "baa".
+                // Then we sort case sensitively, so "aaa" will come before "Aaa".
+                return compareCaseInsensitive(a, b) || compareCaseSensitive(a, b);
+            }
         }
 
-        return a < b ? Comparison.LessThan : Comparison.GreaterThan;
+        function getStringComparerFactory() {
+            // If the host supports Intl, we use it for comparisons using the default locale.
+            if (typeof Intl === "object" && typeof Intl.Collator === "function") {
+                return createIntlCollatorStringComparer;
+            }
+
+            // If the host does not support Intl, we fall back to localeCompare.
+            // localeCompare in Node v0.10 is just an ordinal comparison, so don't use it.
+            if (typeof String.prototype.localeCompare === "function" &&
+                typeof String.prototype.toLocaleUpperCase === "function" &&
+                "a".localeCompare("B") < 0) {
+                return createLocaleCompareStringComparer;
+            }
+
+            // Otherwise, fall back to ordinal comparison:
+            return createFallbackStringComparer;
+        }
+
+        // Hold onto common string comparers. This avoids constantly reallocating comparers during
+        // tests.
+        function createStringComparerCached(locale: CachedLocale, caseInsensitive: boolean) {
+            const cacheKey = locale || "default";
+            const cache = caseInsensitive
+                ? caseInsensitiveCache || (caseInsensitiveCache = {})
+                : caseSensitiveCache || (caseSensitiveCache = {});
+
+            let comparer = cache[cacheKey];
+            if (!comparer) {
+                comparer = createStringComparerNoCache(locale, caseInsensitive);
+                cache[cacheKey] = comparer;
+            }
+
+            return comparer;
+        }
+
+        function createStringComparer(locale: string | undefined, caseInsensitive: boolean) {
+            return locale === undefined || locale === "en-US"
+                ? createStringComparerCached(locale as CachedLocale, caseInsensitive)
+                : createStringComparerNoCache(locale, caseInsensitive);
+        }
+    })();
+
+    let uiCS: Comparer<string> | undefined;

```

Comment:
Nit: use a longer name. Also, might want to put these in an object to ensure they are all set at once.
Also, might as well just set `uiCS` and `uiCI` eagerly at the call to `setUILocale`?

Reply:
I'd rather they were not set at once or created eagerly, I'd prefer these to be created lazily as needed (since they may not be needed).